static/elements/chromedash-app.js

```diff
@@ -0,0 +1,117 @@
+
+
+class ChromedashApp extends LitElement {
+  static get styles() {
+    return [
+      ...SHARED_STYLES,
+      css`
+        #content-component-wrapper[wide] {
+          width: 100%;
+        }
+    `];
+  }
+
+  static get properties() {
+    return {
+      pageComponent: {type: Element},
+      contextLink: {type: String}, // used for the back button in the feature page
+    };
+  }
+
+  constructor() {
+    super();
+    this.pageComponent = null;
+    this.contextLink = '/features';
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+    this.setUpRoutes();
+  }
+
+  setUpRoutes() {
+    page.strict(true); // Be precise about trailing slashes in routes.
+    const header = document.querySelector('chromedash-header');
+
+    // SPA routing rules.  Note that rules are considered in order.
+    // And :var can match any string (including a slash) if there is no slash after it.
+    page('/spa', () => page.redirect('/roadmap'));
+    page('/roadmap', (ctx) => {
+      this.pageComponent = document.createElement('chromedash-roadmap-page');
+      header.currentPage = ctx.path;
+    });
+    page('/myfeatures', (ctx) => {
+      this.pageComponent = document.createElement('chromedash-myfeatures-page');
+      this.contextLink = '/myfeatures';
+      header.currentPage = ctx.path;
+    });
+    page('/features', (ctx) => {
+      this.pageComponent = document.createElement('chromedash-all-features-page');
+      this.contextLink = '/features';
+      header.currentPage = ctx.path;
+    });
+    page('/feature/:featureId', (ctx) => {
+      this.pageComponent = document.createElement('chromedash-feature-page');
+      this.pageComponent.featureId = ctx.params.featureId;
+      this.pageComponent.contextLink = this.contextLink;
+    });
+    page('/settings', (ctx) => {
+      this.pageComponent = document.createElement('chromedash-settings-page');
+      header.currentPage = ctx.path;
+    });
+    page('/metrics/:type/:view', (ctx) => {
+      this.pageComponent = document.createElement('chromedash-stack-rank-page');
+      this.pageComponent.type = ctx.params.type;
+      this.pageComponent.view = ctx.params.view;
+      header.currentPage = ctx.path;
+    });
+    page('/metrics/:type/timeline/:view/:bucketId', (ctx) => {
+      this.pageComponent = document.createElement('chromedash-timeline-page');
+      this.pageComponent.type = ctx.params.type;
+      this.pageComponent.view = ctx.params.view;
+      this.pageComponent.selectedBucketId = ctx.params.bucketId;
+      header.currentPage = ctx.path;
+    });
+
+    page.start();
+  }
+
+  render() {
+    // TODO: Create precomiled main css file and import it instead of inlining it here
+    return html`
+      <link rel="stylesheet" href="/static/css/main.css">
+      <div id="app-content-container">
+        <div>
+          <div class="main-toolbar">
+            <div class="toolbar-content">
+              <slot name="header"></slot>
```

Comment:
Why use slots instead of putting `<chromedash-header ...>` here?

Slots can keep a component reusable, but this is our top-level app, it is not going to be reused in another context.

Reply:
It's because of the Google sign-in button. The current `<chromedash-header>` component insert it to its light DOM to allow the Google Auth library to find it. The `<chromedash-header>`'s light DOM is at document level in the MPA version, which works fine.

But in the SPA version, if `<chromedash-header>` is within `<chromedash-app>`,  `<chromedash-header>`'s light DOM would be in the `<chromedash-app>`'s shadow DOM, which blocks the Auth library's access. That's why I used the slot to make sure the header component is at the document level.

I'd love to not use the slot and there are some things I want to try. But I couldn't make any meaningful changes to the `<chromedas-header>` now because it's still used for the MPA.

Reply:
This is actually resolved after my changes.